spring-messaging/src/main/java/org/springframework/messaging/core/GenericMessagingTemplate.java

```diff
@@ -142,20 +212,25 @@ protected final void doSend(MessageChannel channel, Message<?> message) {
 		Object originalReplyChannelHeader = requestMessage.getHeaders().getReplyChannel();
 		Object originalErrorChannelHeader = requestMessage.getHeaders().getErrorChannel();
 
-		TemporaryReplyChannel tempReplyChannel = new TemporaryReplyChannel();
-		requestMessage = MessageBuilder.fromMessage(requestMessage).setReplyChannel(tempReplyChannel).
-				setErrorChannel(tempReplyChannel).build();
+		long sendTimeout = sendTimeout(requestMessage);
+		long receiveTimeout = receiveTimeout(requestMessage);
+
+		TemporaryReplyChannel tempReplyChannel = new TemporaryReplyChannel(this.throwExceptionOnLateReply);
+		requestMessage = MessageBuilder.fromMessage(requestMessage).setReplyChannel(tempReplyChannel)
+				.setHeader(this.sendTimeoutHeader, null)
+				.setHeader(this.receiveTimeoutHeader, null)
+				.setErrorChannel(tempReplyChannel).build();
 
 		try {
-			doSend(channel, requestMessage);
+			doSend(channel, requestMessage, sendTimeout);
 		}
 		catch (RuntimeException ex) {
 			tempReplyChannel.setSendFailed(true);
 			throw ex;
 		}
 
-		Message<?> replyMessage = this.doReceive(tempReplyChannel);
-		if (replyMessage != null) {
+		Message<?> replyMessage = this.doReceive(tempReplyChannel, receiveTimeout);
+		if (replyMessage != null && (originalReplyChannelHeader!= null || originalErrorChannelHeader != null)) {
```

Comment:
Sorry - I just issued a new PR on that branch (having first reset it to master).

I issued another PR: https://github.com/spring-projects/spring-framework/pull/1440

Thanks